### PR TITLE
Raykoaza/issue/3009 shuf depen

### PR DESCRIPTION
**Pull request by raykoaza**
Added shuf/gshuf(macos) dependency check for shuf package on Linux (gshuf on mac).

https://github.com/skypilot-org/skypilot/issues/3009

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

Built and executed the code, now it's exiting if the command is not installed rather than failing with "command not found error"

### CODE REVIEW

File: sky/templates/kubernetes-port-forward-proxy-command.sh.j2 [hunk 1]
```diff
@@ -34,7 +34,7 @@ local_port=$(awk -F: '/Forwarding from 127.0.0.1/{print $2}' "${KUBECTL_OUTPUT}"
 # Clean up the temporary file
 rm -f "${KUBECTL_OUTPUT}"
 
-# Add handler to terminate the port-forward process when this script exits.
+Added shuf/gshuf(macos) dependency check# Add handler to terminate the port-forward process when this script exits.
```
Comment (romilbhardwaj) [inline sky/templates/kubernetes-port-forward-proxy-command.sh.j2:37]: Unintended change?

File: sky/templates/kubernetes-port-forward-proxy-command.sh.j2 [hunk 2]
```diff
@@ -45,8 +45,26 @@ done
...
+      fi
+  else
+      # Check if shuf is installed
+      if [[ $(command -v shuf) == "" ]]; then
+          echo "shuf is not installed. Please install shuf first."
+          exit 1
+      fi
+      echo "shuf is installed."
+  fi
+fi
```
Comment (romilbhardwaj) [inline sky/templates/kubernetes-port-forward-proxy-command.sh.j2:67]: For consistency, we should move these checks to after L14, like we have for `socat` and `nc`. We should also probably detect the right binary (`gshuf` vs `shuf`) during that check and use that later on here at L48.

File: sky/templates/kubernetes-port-forward-proxy-command.sh.j2 [hunk 3]
```diff
@@ -45,8 +45,26 @@ done
 # To avoid errors when many concurrent requests are sent (see https://github.com/skypilot-org/skypilot/issues/2628),
 # we add a random delay before establishing the socat connection.
 # Empirically, this needs to be at least 1 second. We set this to be random between 1 and 2 seconds.
-sleep $(shuf -i 10-20 -n 1 | awk '{printf "%f", $1/10}')
-
+if which shuf > /dev/null; then
+  sleep $(shuf -i 10-20 -n 1 | awk '{printf "%f", $1/10}')
```
Comment (romilbhardwaj) [inline sky/templates/kubernetes-port-forward-proxy-command.sh.j2:49]: We need to add shuf/gshuf checks to `kubernetes_utils.py::check_port_forward_mode_dependencies()` as well, and include installation commands there.

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @raykoaza! Left some comments.

File: sky/templates/kubernetes-port-forward-proxy-command.sh.j2 (hunk 1 above)
Comment (raykoaza) [reply sky/templates/kubernetes-port-forward-proxy-command.sh.j2:37]: Yes, typo. Will remove this.

File: sky/templates/kubernetes-port-forward-proxy-command.sh.j2 [hunk 4]
```diff
@@ -13,6 +13,25 @@ if ! command -v nc > /dev/null; then
   exit
 fi
 
+# Checks if shuf is installed(Linux/macos)
+if [[ "$OSTYPE" == "darwin"* ]]; then
```
Comment (romilbhardwaj) [inline sky/templates/kubernetes-port-forward-proxy-command.sh.j2:17]: We probably don't want to have separate checks for os and then check only for gshuf. For example, my mac machine has shuf available.

Can we simplify this check to:
1. Check if `shuf` is available
2. If not, check if `gshuf` is available
3. Else, raise error.

File: sky/utils/kubernetes_utils.py [hunk 5]
```diff
@@ -1016,12 +1017,17 @@ def fill_ssh_jump_template(ssh_key_secret: str, ssh_jump_image: str,
...
     # We store the dependency list as a list of lists. Each inner list
     # contains the name of the dependency, the command to check if it is
     # installed, and the package name to install it.
+    osused = pltsys()
     dependency_list = [['socat', ['socat', '-V'], 'socat'],
                        ['nc', ['nc', '-h'], 'netcat']]
+    if osused == 'Darwin':
+        dependency_list.append(['gshuf', ['gshuf', '--version'], 'coreutils'])
+    elif osused == 'Linux':
+        dependency_list.append(['shuf', ['shuf', '--version'], 'coreutils'])
```
Comment (romilbhardwaj) [inline sky/utils/kubernetes_utils.py:1030]: Similar to above, can we first check for shuf, then fallback to gshuf if its not present? That way we don't need OS-specific checks.

Comment (raykoaza) [reply sky/utils/kubernetes_utils.py:1030]: In that case dependency list can have only shuf and not bother about gshuf? I had it this way caz, in future if we have any specific package differences between each OS, we can add it to this block.

**Comment by raykoaza** [comment]
@romilbhardwaj , I see remote and local are in sync. Why it's giving conflict warning. do you see any issue?

**Review by raykoaza — COMMENTED** [review]
make sense, as package name is already part of the return statement

**Pull request closed without merging**